exporter/exporterhelper/persistent_queue.go

```diff
@@ -0,0 +1,343 @@
+//      │              │
+//    write          read
+//    index          index
+//
+type persistentContiguousStorage struct {
+	logger        *zap.Logger
+	client        storage.Client
+	mu            sync.Mutex
+	queueName     string
+	readIndex     uint64
+	writeIndex    uint64
+	readIndexKey  string
+	writeIndexKey string
+	retryDelay    time.Duration
+	unmarshaler   requestUnmarshaler
+	reqChan       chan request
+	stopped       *atomic.Uint32
+}
+
+const (
+	queueNameKey      = "queueName"
+	zapItemKey        = "itemKey"
+	itemKeyTemplate   = "it-%d"
+	readIndexKey      = "ri"
+	writeIndexKey     = "wi"
+	defaultRetryDelay = 500 * time.Millisecond
+)
+
+// newPersistentQueue creates a new queue backed by file storage
+func newPersistentQueue(ctx context.Context, name string, capacity int, logger *zap.Logger, client storage.Client, unmarshaler requestUnmarshaler) *persistentQueue {
+	return &persistentQueue{
+		logger:   logger,
+		exitChan: make(chan int),
+		capacity: capacity,
+		storage:  newPersistentContiguousStorage(ctx, name, logger, client, unmarshaler),
+	}
+}
+
+// StartConsumers starts the given number of consumers which will be consuming items
+func (pq *persistentQueue) StartConsumers(num int, callback func(item interface{})) {
+	pq.numWorkers = num
+	var startWG sync.WaitGroup
+
+	factory := func() queue.Consumer {
+		return queue.ConsumerFunc(callback)
+	}
+
+	for i := 0; i < pq.numWorkers; i++ {
+		pq.stopWG.Add(1)
+		startWG.Add(1)
+		go func() {
+			startWG.Done()
+			defer pq.stopWG.Done()
+			consumer := factory()
+
+			for {
+				select {
+				case req := <-pq.storage.get():
+					consumer.Consume(req)
+				case <-pq.exitChan:
+					return
+				}
+			}
+		}()
+	}
+	startWG.Wait()
+}
+
+// Produce adds an item to the queue and returns true if it was accepted
+func (pq *persistentQueue) Produce(item interface{}) bool {
+	pq.produceMu.Lock()
+	defer pq.produceMu.Unlock()
+
+	if pq.storage.size() >= pq.capacity {
+		return false
+	}
+	err := pq.storage.put(item.(request))
+	return err == nil
+}
+
+// Stop stops accepting items, shuts down the queue and closes the persistent queue
+func (pq *persistentQueue) Stop() {
+	pq.storage.stop()
+	close(pq.exitChan)
+	pq.stopWG.Wait()
+}
+
+// Size returns the current depth of the queue
+func (pq *persistentQueue) Size() int {
+	return pq.storage.size()
+}
+
+// newPersistentContiguousStorage creates a new file-storage extension backed queue. It needs to be initialized separately
+func newPersistentContiguousStorage(ctx context.Context, queueName string, logger *zap.Logger, client storage.Client, unmarshaler requestUnmarshaler) *persistentContiguousStorage {
+	wcs := &persistentContiguousStorage{
+		logger:      logger,
+		client:      client,
+		queueName:   queueName,
+		unmarshaler: unmarshaler,
+		reqChan:     make(chan request),
+		stopped:     atomic.NewUint32(0),
+		retryDelay:  defaultRetryDelay,
+	}
+	initPersistentContiguousStorage(ctx, wcs)
+	go wcs.loop(ctx)
```

Comment:
I wouldn't personally use context in such a setting. Context is meant to be used as a short lived cancellation mechanism. In this setting it will live as long as `loop` doesn't exit whereas `newPersistentContiguousStorage` caller should expect that the context that he passes in here is used only for function calls right here, right now, not in a separate goroutine, possibly outliving the caller.